ecl/hql/hqlgram2.cpp

```diff
@@ -12329,30 +12330,28 @@ void parseAttribute(IHqlScope * scope, IFileContents * contents, HqlLookupContex
         throw;
     }
     OwnedHqlExpr parsed = scope->lookupSymbol(name, LSFsharedOK|LSFnoreport, ctx);
```

Comment:
I think it would be useful to have a comment at the start of parseAttribute which details how the flags are meant to work:
E.g.

// If ctx.hasCacheLocation() then the cache entry is always updated with the dependency information (unless it is unavailable i.e. a forward scope)
// If ctx.ignoreSimplified() then the expressions are never simplified and the parsing behaves the same way as 6.x
// if ctx.ignoreCache() then the contents of the cache are not used, but the cache is still updated
// if ctx.regenerateCache() then the contents of the cache are not used and the cache is forced to be updated
// if ctx.checkSimpleDef() then simplified definitions are created and checked - even if there is no cache configured.

I think that covers all the cases.  That was helpful for me to spell it out - I can see one conflict.  If ignoreSimplified is true should the cache information be updated?  If it is not the dependencies will not be correct, if it is the simplified will be forced to be blank.  I the latter is probably better since it means the dependency cache will still work if the user has disabled the simplified expressions.

Also, it highlights a significant optimization.  If the cache was up to date but was not simplified, then there is no benefit in creating a simplified expression - it failed last time, and will fail again.

There is a related issue for the expression too complex case, but I'll comment on that PR separately.

Reply:
When ignoreCache() is enabled, should expressions be simplified and used?